Core/Src/internalResistance.c

```diff
@@ -9,128 +9,213 @@
+                        it is filled with a placeholder value to indicate it cannot be used for calculation */
+                    voltageAvgBuffer[i][j][avgBufferIndex] = -1000.0f;
+
+                    // Static data good var is reset for the next discrete buffer filling cycle
+                    voltageDataGood[i][j] = true;
                 }
-            }
-            else
-            {
-                currentAvgBuffer[avgBufferIndex] = -1000.0f;
-                dataGood = true;
-            }
+            }    
+        }
+    }
+    else
+    {
+        /*  If bad data was in the corresponding discrete voltage buffer, 
+            it is filled with a placeholder value to indicate it cannot be used for calculation */
+        currentAvgBuffer[avgBufferIndex] = -1000.0f;
+
+        // Static data good var is reset for the next discrete buffer filling cycle
+        currentDataGood = true;
+    }
+
+    // Cycle the circular average buffers and loop if necessary
+    if (++avgBufferIndex >= AVERAGE_BUFFER_SIZE)
+    {
+        avgBufferIndex = 0;
+    }
 
-            avgBufferIndex++;
+    // Max and min data initilized to far left and right values
+    float maxCurrent = -1000.0f;
+    float minCurrent = 1000.0f;
 
-            if (avgBufferIndex >= 10)
+    // Default max and min index is 0
+    uint32_t maxCurrentIndex = 0;
+    uint32_t minCurrentIndex = 0;
+
+    // Calculate max and min currents in buffer
+    for(int32_t i = 0; i < AVERAGE_BUFFER_SIZE; i++)
+    {
+        // If average current buffer containes placeholder value, -1000.0f, it is excluded from the max/min calc
+        if(currentAvgBuffer[i] > -900.0f)
+        {
+            if(currentAvgBuffer[i] > maxCurrent)
+            {
+                maxCurrentIndex = i;
+            }
+            else if(currentAvgBuffer[i] < minCurrent)
             {
-                avgBufferIndex = 0;
+                minCurrentIndex = i;
             }
+        }
+    }
 
-            float maxCurrent = -1000.0f;
-            float minCurrent = 1000.0f;
-            uint32_t maxCurrentIndex = 0;
-            uint32_t minCurrentIndex = 0;
+    //  Calculate greatest dI and dV contained in buffer
+    /*  If avg buffer saturated with -1000.0f placeholder values, default index 0 will still be set, 
+        and delta current will equal 0 (-1000.0f - -1000.0f)    */
+    float deltaCurrent = currentAvgBuffer[maxCurrentIndex] - currentAvgBuffer[minCurrentIndex];
 
-            // Calculate max and min currents in buffer
-            for(int32_t i = 0; i < 10; i++)
+    // If delta current threshold is met, internal resistance can be calulated and updated
+    // This threshold should always exceed 0, so if avg buffer is saturated with -1000.0f, operation will be skipped
+    if((fabs(deltaCurrent) >= IR_CALC_MIN_CURRENT_DELTA) && (fabs(deltaCurrent) <= CURRENT_LOW_TO_HIGH_SWITCH_THRESHOLD))
+    {
```

Comment:
This doesn't verify that current sensor channel 1 is being used. Part of me wants to say, lets just not care and let dI be as big as it can be, even if we lose a little accuracy.